### ISSUE TITLE
Capture diagnostics for intermittent CAMP GUI hangs (Qt heartbeat + backtrace-on-stall) — root cause unconfirmed

### ISSUE THREAD

**Issue by rolker** [rolker/AutonomousMissionPlanner#52]
## Why

Two operator-reported **CAMP-unresponsive events** during the 2026-05-22 BizzyBoat deployment (analysis: `rolker/unh_echoboats_project11#166`). **Root cause is unconfirmed** — and importantly, a GUI hang **leaves no trace in the logs**, while the hard log evidence we *do* have is explained by mundane causes:
- Event 1's CAMP `SIGKILL` was the launch's normal SIGINT→SIGTERM→SIGKILL shutdown escalation after the operator restarted the stack — and ~6 other nodes needed SIGKILL in the same shutdown due to an **rmw_zenoh "close operation timed out"** hang. So it is *not* evidence of a freeze.
- Event 2 (camp alive, GUI hung, ignored SIGTERM, solo `pkill` while the rest of the stack survived) is the only event whose signature resists a mundane explanation — but it's reconstructed from an agent-written field log, not a captured artifact.

So we cannot confirm CAMP froze, let alone why. **This issue is about capturing the next occurrence definitively**, not implementing a fix for an unproven cause.

## Hypothesis (unproven — to be confirmed or refuted)

GUI hang under a **link-recovery data burst**. Suspect path: CAMP runs a `rclcpp::executors::MultiThreadedExecutor` on a dedicated thread (`src/camp/ros/node_thread.cpp`) and hands marker data to the Qt thread via `shared_ptr` snapshots (`src/camp/markers/markers.h`). A MarkerArray flood (boat-side bag showed `path_follower_visualization` 0→253/min at the Event-2 timestamp, on top of a both-legs link-recovery backlog) could contend/saturate that handoff and wedge the GUI thread.

## Deliverables (capture-first)

1. **Qt-timer GUI heartbeat** — publish a heartbeat from a `QTimer` on the GUI thread. If it stalls *while the ROS executor thread is still alive*, that is a **definitive "GUI-thread hung"** signature (and distinguishes it from a full-process wedge or a clean shutdown).
2. **Auto-backtrace watchdog** — on heartbeat stall, dump all-thread backtraces (`gdb -p`/`eu-stack`/`gstack`) to a file **before** the operator kills CAMP. Turns the next freeze into proof with no operator action.
3. **Operator one-liner** (documented, for manual capture if the watchdog isn't in yet):
   ```
   gdb -p $(pgrep -f CCOMAutonomous) -batch -ex "thread apply all bt" > /tmp/camp_freeze_$(date +%s).txt 2>&1
   ```
4. **(Optional) bench repro** — replay the operator/boat bags (or a synthetic marker+diagnostic flood) into CAMP over a `tc netem` degraded/recovering link; if it wedges, attach gdb.
5. **Code audit** of the executor↔Qt marker-snapshot path for blocking calls / lock-ordering that could deadlock under burst (supporting evidence to match against any captured backtrace).

## Acceptance

The next CAMP hang yields a thread backtrace that identifies the stuck path — **confirming or refuting** the executor↔Qt-handoff hypothesis. Only then does a fix get scoped.

Part of `rolker/unh_echoboats_project11#166`.

---
**Authored-By**: `Claude Code Agent`
**Model**: `Claude Opus 4.7 (1M context)`

**Comment by rolker** [comment]
**Teardown-fragility data point (from the #166 dive):** CAMP also **segfaulted (SIGSEGV, exit -11) on normal shutdown** at the end of the 2026-05-22 mission (20:27), distinct from the GUI-hang events. It's in the same Qt + `MultiThreadedExecutor` teardown neighborhood — worth a glance while implementing the heartbeat watchdog, since a clean-shutdown path and the watchdog touch the same lifecycle code. Not the hang, but a sign the teardown path isn't robust.

---
**Authored-By**: `Claude Code Agent`
**Model**: `Claude Opus 4.7 (1M context)`

**Comment by rolker** [comment]
## 3rd occurrence — 2026-05-29 BizzyBoat deployment (unh_echoboats_project11#197), salmon

Another intermittent hang, again **no definitive capture** (deliverables 1–2 not yet built — reinforces this issue). Field detail from the salmon log:

- `CCOMAutonomousMissionPlanner` PID 138488 (launched 12:14:42), **SIGKILL (exit -9) at 15:22:29 EDT**, restarted 15:22:34. Operator force-killed (clean signal wouldn't take) → consistent with unresponsive.
- **Last GUI activity 15:15:43**; operator noticed ~15:19 — a **~4-min gap**. Operator notes the freeze may have happened earlier and interaction (panning / right-click context-menu for goto/hover) was *when it was noticed*, not necessarily the trigger. Recurring — seen ≥2× before.
- Last log line before the kill: `Message Filter dropping message: frame 'bizzy/base_link' ... 'timestamp ... earlier than all the data in the transform cache'`. The **same `bizzy/base_link` TF cache miss reappeared on the restarted instance at 15:23:52** → ongoing TF/clock state, not a one-shot.
- Zenoh anomalies around the window: 15:02:59 `rmw_zenoh SubscriberCallback triggered over 0/bizzy/marine/heartbeat`; post-restart `Scouting delay elapsed` + `Didn't receive DeclareFinal for interest ... Timeout(10s)`.
- Negative: **no OOM** (dmesg clean), no `journalctl --user` CAMP entries.

### New angle for the hypothesis set
Alongside the executor↔Qt marker-burst hypothesis, this points at a **TF message-filter path**: a persistent `bizzy/base_link` TF cache miss (possible salmon↔gabby clock skew) could stall a Qt-thread-bound TF wait / message-filter callback, with flaky Zenoh discovery compounding. Probes when quiet: `tf2_echo bizzy/map bizzy/base_link` stability, `ros2 topic hz /bizzy/tf`, and `chronyc tracking` salmon vs gabby for clock drift.

Doesn't change the capture-first plan — the Qt heartbeat + backtrace-on-stall (deliverables 1–2) would have turned this into proof too.

Part of unh_echoboats_project11#197.

---
**Authored-By**: `Claude Code Agent`
**Model**: `Claude Opus 4.8 (1M context)`

**Comment by rolker** [comment]
Deployment #205 (2026-06-02, salmon operator station) captured diagnostics for an intermittent **partial** freeze that may inform this:

- **Selective freeze:** heartbeat + GPS-position displays froze while the collision-avoidance polygons kept moving — not a full GUI hang. Steady upstream data (`/bizzy/marine/heartbeat` 1.00 Hz clean) while the display froze → a consumer/render problem, not dropped comms (payload loss = 0 throughout).
- **Ruled out:** tide-TF stall (`map→chart_datum` steady 1 Hz, 0 gaps >1.6 s over 140 s).
- **Biggest absolute consumer:** `/bizzy/local_costmap/costmap_windowed` = 640 KB/msg × 1.36 Hz ≈ 950 KB/s (800×800 @ 0.25 m), ~50× all other CAMP topics combined — a plausible Qt-render-thread starver.
- **But (operator note):** that costmap was the same size on prior days **without** this freeze, so it's unlikely to be *the* cause today. What changed this run is the flipped avoider plugin (rolker/unh_marine_navigation#206) adding/altering viz topics & message volume to CAMP. The costmap is the biggest *absolute* load, not necessarily the *delta* — **needs a CAMP-subscription before/after comparison** to separate them.
- Possible robustness fix regardless: rasterize the OccupancyGrid off the Qt main thread so a big grid can't starve other displays.

Deployment: rolker/unh_echoboats_project11#205.

---
**Authored-By**: `Claude Code Agent`
**Model**: `Claude Opus 4.8 (1M context)`

**Comment by rolker** [comment]
**New occurrence (2026-06-05 pier outing).** CAMP locked up again during on-water
ops. Adding as a dated data point for this intermittent-hang investigation.

Durable operator note (salmon bag sidecar
`~/data/logs/operator/2026-06-05/operator_log.jsonl`):

> **14:31:20 EDT** — "camp locked up"  (boat recovered ~14:46:35 EDT)

No backtrace / Qt-heartbeat capture was taken at the time. Distinct from the
2026-06-03 trackline-removal **crash** (#65, which auto-respawned) — this was a
**hang/lockup**, matching the failure class tracked here and in
rolker/unh_echoboats_project11#166.

Context: the 2026-06-03 run included a CAMP callback-group concurrency rework
(merged to `jazzy`), so the diagnostics-on-stall tooling requested in this issue
is the right next step to tell whether the GUI thread is blocked vs. the executor.

---
**Authored-By**: `Claude Code Agent`
**Model**: `Claude Opus 4.8 (1M context)`
